### PR TITLE
Provide a nicer image build failed message

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

Make the message you see when an image build fails nicer for humans, including the recommendation for `gp validate`. 

## Related Issue(s)

Fixes ENT-724

## How to test

1. Open https://ft-nicer-ie500177c4c.preview.gitpod-dev.com/new#https://github.com/kylos101/gitpod-custom-image/tree/kylos101/build-fail

### CODE REVIEW

**Comment by kylos101** [comment]
Hey @filiptronicek , to test you could try using an image build that will fail, like this one: https://github.com/kylos101/gitpod-custom-image/tree/kylos101/build-fail

It fails like so in catfood:
image

**Comment by filiptronicek** [comment]
@kylos101 thanks for sharing, I have been able to get this screen as well. What I was originally curious about is this image build screen without any logs, where the improved error message would be a very nice addition:
image

For the purposes of testing this PR, your suggestion will be enough, though. Thanks!

**Comment by kylos101** [comment]
Oh, I see - have you tried deleting the image build pod, before logs start to stream? We might be able to enable debug with image-builder-mk3 too.  I think that introduces a delay in the image build process, which might give us time to delete the imagebuild pod itself.

**Comment by mustard-mh** [comment]
@filiptronicek maybe you're looking for this one https://github.com/gitpod-io/gitpod/blob/359934c65742b641d8582dac0d92613e04462d06/components/supervisor/pkg/supervisor/supervisor.go#L1603-L1604 ?

content inside `/dev/termination-log` will be consider as failed reason / status of a pod

**Comment by filiptronicek** [comment]
Thanks so much for sharing that, @mustard-mh, that's exactly what I was looking for (I originally thought we display the error we catch from the workspace starter).

> update: but it's not always work well, like [this one](https://github.com/gitpod-io/gitpod/blob/359934c65742b641d8582dac0d92613e04462d06/components/supervisor/pkg/supervisor/supervisor.go#L432-L434), it will show another message relates to pod sometimes if I remember it correctly

Oh, that's a bummer. Do we have an issue for that? Seems like we should to make our error messages reliable

**Comment by mustard-mh** [comment]
> Oh, that's a bummer. Do we have an issue for that? Feels like we should to make our error messages reliable

@filiptronicek Searched, seems we don't have such issue

**Comment by mustard-mh** [comment]
@filiptronicek ws-manager will read content of `/dev/termination-log` here https://github.com/gitpod-io/gitpod/blob/fea7d78cf34b73101251db8f550a4f66d23c6171/components/ws-manager-mk2/controllers/status.go#L140

In case you need to investigate further

**Comment by filiptronicek** [comment]
@mustard-mh could you help me understand why we fail the image build as part of the supervisor's `stopWhenTasksAreDone`? I never thought of that phase as a task because it can't run in parallel with the other ones. Are we adding it somewhere artificially to the list? 

If so, we could make this work by communicating through the success channel not just the success state, but also some task identifier. If that matched the image build, we could trigger a special message.

**Comment by filiptronicek** [comment]
🎉

image

**Comment by lucasvaltl** [comment]
> 🎉
> 
> image

Regarding the actual error message we show. Is my assumption correct that in effectively every case this is an error caused by the user? I.e. something is wrong with the image being built, e.g. a mistake in the docker file? If this is the case, let's make this clear in the error message we show. E.g. (feel free to improve with AI) "The image failed to build. Most likely, this is due to a misconfiguration of the image  that causes an error due to the build process. To debug this, it is advised to use `gp validate` (link to docs)."

**Comment by filiptronicek** [comment]
> @lucasvaltl: Regarding the actual error message we show. Is my assumption correct that in effectively every case this is an error caused by the user?

Yes, when a headless task fails, we categorize it as a user error.

I'm going to try with another message, adapted from your suggestion. Trying to keep it short to fit in all places we show these messages (wish we had a link shortener for the long `gp validate` docs link 🙈)

**Comment by filiptronicek** [comment]
Argh, it doesn't really fit

image

File: components/supervisor/pkg/supervisor/config.go [hunk 1]
```diff
@@ -468,6 +468,11 @@ func (c WorkspaceConfig) isDebugWorkspace() bool {
 	return c.DebugWorkspaceType != api.DebugWorkspaceType_noDebug
 }
 
+// isImageBuild returns true if the workspace is an image build.
+func (c WorkspaceConfig) isImageBuild() bool {
+	return os.Getenv("BOB_DOCKERFILE_PATH") != ""
```
Comment (mustard-mh) [inline components/supervisor/pkg/supervisor/config.go:473]: nit: Other config will [read env when init](https://github.com/gitpod-io/gitpod/blob/e266df1bc80b7f887cadb294f402ed6eec37a07b/components/supervisor/pkg/supervisor/config.go#L310), we should do it the same for isImageBuild to align with others

image

**Review by mustard-mh — APPROVED** [review]
LGTM

|Image Build|
|:-:|
|image|

nit: see screenshot, the links we provide on 'Learn More' and logs are different.

**Review by geropl — APPROVED** [review]

File: components/supervisor/pkg/supervisor/config.go (hunk 1 above)
Comment (filiptronicek) [reply components/supervisor/pkg/supervisor/config.go]: Thanks for pointing this out, fixed in [`e1b1d79` (#20281)](https://github.com/gitpod-io/gitpod/pull/20281/commits/e1b1d794305bd66e05a93b2b8c492a069ec68d19).

**Pull request merged**